src/NHibernate/Cache/FilterKey.cs

```diff
@@ -8,10 +10,16 @@
 namespace NHibernate.Cache
 {
 	[Serializable]
-	public class FilterKey
+	public class FilterKey : IDeserializationCallback
 	{
 		private readonly string _filterName;
-		private readonly Dictionary<string, TypedValue> _filterParameters = new Dictionary<string, TypedValue>();
+
+		// Sets and dictionaries are populated last during deserialization, causing them to be potentially empty
+		// during the deserialization callback. This causes them to be unreliable when used in hashcode computations.
+		// So better serialize them as other structures and use those other structures in hashcode computation.
+		[NonSerialized]
+		private Dictionary<string, TypedValue> _filterParameters = new Dictionary<string, TypedValue>();
+		private IEnumerable<KeyValuePair<string, TypedValue>> _serializedFilterParameters;
```

Comment:
What is the point declare it as IEnumerable?

Reply:
This avoids a cast in `GetHashCode` and `Equals` implementations, when doing things like `var filterParameters = _serializedFilterParameters ?? _filterParameters;`.